File: leetcode/c++/medium/_2661_first_completely_painted_row_or_column.cpp
// Leetcode Problem 2661

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    int firstCompleteIndex(vector<int>& arr, vector<vector<int>>& mat) {
        int m = mat.size();
        int n = mat[0].size();

        vector<int> frequency_column(m, 0);
        vector<int> frequency_row(n, 0);
        vector<tuple<int, int>> positions(m*n, make_tuple(0, 0));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                positions[mat[i][j] - 1] = make_tuple(i, j);
            }
        }

        for (int pos = 0; pos < arr.size(); pos++) {
            tuple<int, int> position = positions[arr[pos] - 1];
            
            frequency_column[get<0>(position)]++;
            frequency_row[get<1>(position)]++;

            if (frequency_column[get<0>(position)] == n || frequency_row[get<1>(position)] == m) {
                return pos;
            }
        }

        return arr.size() - 1;
    }
};


class TestFirstCompleteIndex : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestFirstCompleteIndex, SimpleTestCase1) {
    vector<int> arr = {1, 3, 4, 2};
    vector<vector<int>> mat = {
        {1, 4},
        {2, 3}
    };

    EXPECT_EQ(solution.firstCompleteIndex(arr, mat), 2);
}

TEST_F(TestFirstCompleteIndex, SimpleTestCase2) {
    vector<int> arr = {2, 8, 7, 4, 1, 3, 5, 6, 9};
    vector<vector<int>> mat = {
        {3, 2, 5},
        {1, 4, 6},
        {8, 7, 9}
    };

    EXPECT_EQ(solution.firstCompleteIndex(arr, mat), 3);
}

TEST_F(TestFirstCompleteIndex, SimpleTestCase3) {
    vector<int> arr = {1, 4, 5, 2, 6, 3};
    vector<vector<int>> mat = {
        {4, 3, 5},
        {1, 2, 6}
    };

    EXPECT_EQ(solution.firstCompleteIndex(arr, mat), 1);
}

TEST_F(TestFirstCompleteIndex, SimpleTestCase4) {
    vector<int> arr = {6, 2, 3, 1, 4, 5};
    vector<vector<int>> mat = {
        {5, 1},
        {2, 4},
        {6, 3}
    };

    EXPECT_EQ(solution.firstCompleteIndex(arr, mat), 2);
}

File: leetcode/rust/problems/src/medium/_2661_first_completely_painted_row_or_column.rs
// Leetcode Problem 2661

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn first_complete_index(arr: Vec<i32>, mat: Vec<Vec<i32>>) -> i32 {
        let m = mat.len();
        let n = mat[0].len();

        let mut frequency_row = vec![0; n];
        let mut frequency_column = vec![0; m];
        let mut positions: Vec<(usize, usize)> = vec![(0, 0); m*n];

        for i in 0..m {
            for j in 0..n {
                positions[mat[i][j] as usize - 1] = (i, j);
            }
        }

        for (pos, num) in arr.iter().enumerate() {
            let position = positions[*num as usize - 1];

            frequency_column[position.0] += 1;
            frequency_row[position.1] += 1;

            if frequency_row[position.1] == m || frequency_column[position.0] == n {
                return pos as i32;
            }
        }

        (arr.len() - 1) as i32
    }
}


#[test]
fn test_first_complete_index() {
    let arr1 = vec![1, 3, 4, 2];
    let mat1 = vec![
        vec![1, 4],
        vec![2, 3]
    ];

    let arr2 = vec![2, 8, 7, 4, 1, 3, 5, 6, 9];
    let mat2 = vec![
        vec![3, 2, 5],
        vec![1, 4, 6],
        vec![8, 7, 9]
    ];

    let arr3 = vec![1, 4, 5, 2, 6, 3];
    let mat3 = vec![
        vec![4, 3, 5],
        vec![1, 2, 6]
    ];

    let arr4 = vec![6, 2, 3, 1, 4, 5];
    let mat4 = vec![
        vec![5, 1],
        vec![2, 4],
        vec![6, 3]
    ];

    assert_eq!(Solution::first_complete_index(arr1, mat1), 2);
    assert_eq!(Solution::first_complete_index(arr2, mat2), 3);
    assert_eq!(Solution::first_complete_index(arr3, mat3), 1);
    assert_eq!(Solution::first_complete_index(arr4, mat4), 2);
}